src/client/Microsoft.Identity.Client/Internal/IdToken.cs

```diff
@@ -79,6 +80,28 @@ public string GetUniqueId()
 
         public ClaimsPrincipal ClaimsPrincipal { get; private set; }
 
+
+        private static IdToken ClaimsToToken(List<Claim> claims)
+        {
+            var principal = new ClaimsPrincipal(new ClaimsIdentity(claims));
+            return new IdToken
+            {
+                ClaimsPrincipal = principal,
+                ObjectId = FindClaim(claims, IdTokenClaim.ObjectId),
+                Subject = FindClaim(claims, IdTokenClaim.Subject),
+                TenantId = FindClaim(claims, IdTokenClaim.TenantId),
+                PreferredUsername = FindClaim(claims, IdTokenClaim.PreferredUsername),
+                Name = FindClaim(claims, IdTokenClaim.Name),
+                Email = FindClaim(claims, IdTokenClaim.Email),
+                Upn = FindClaim(claims, IdTokenClaim.Upn),
+                GivenName = FindClaim(claims, IdTokenClaim.GivenName),
+                FamilyName = FindClaim(claims, IdTokenClaim.FamilyName)
+            };
+
+            static string FindClaim(List<Claim> claims, string type) =>
+                claims.SingleOrDefault(_ => string.Equals(_.Type, type, StringComparison.OrdinalIgnoreCase))?.Value;
```

Comment:
@bgavrilMS do u want a SingleOrDefault or a FirstOrDefault

i guess my real question is hos should duplicate claim types be handled when parsing?

Reply:
Good question. An ID Token must be a JWT token https://openid.net/specs/openid-connect-core-1_0.html#IDToken and in a JWT token the claim names must be unique https://www.rfc-editor.org/rfc/rfc7519.html#page-8

So I think `SingleOrDefault` is appropriate, i.e. it's better to fail in case a duplicate claim exists.